cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright © 2015 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.internal.app.runtime.workflow;
+
+import co.cask.cdap.api.workflow.WorkflowToken;
+import com.google.common.collect.Maps;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Implementation of the {@link WorkflowToken} interface.
+ */
+public class BasicWorkflowToken implements WorkflowToken {
+  private Map<String, Map<String, Long>> mapReduceCounters = Maps.newHashMap();
+
+  @Override
+  public Map<String, Map<String, Long>> getMapReduceCounters() {
+    return mapReduceCounters;
+  }
+
+  public void setMapReduceCounters(Map<String, Map<String, Long>> mapReduceCounters) {
+    this.mapReduceCounters = mapReduceCounters;
```

Comment:
you should make a defensive copy into an Immutable map here, because you can't know whether the caller will modify the map later-on.